### PR TITLE
Feature/cbi 2011 fallback mechanism ga integration

### PR DESCRIPTION
**Pull request by DmitriyMolch**
https://cerenetwork.atlassian.net/browse/CBI-2011

### CODE REVIEW

File: config/chainbridge-runtime-config.cere.dev.js [hunk 1]
```diff
@@ -4,6 +4,10 @@ const cereTokenDecimals = 10;
 
 window.__RUNTIME_CONFIG__ = {
   CHAINBRIDGE: {
+    ga: {
+      trackingId: "G-EVYKNM010Y",
+      appName: "chainbridge",
```
Comment (AndreiNavoichyk) [inline config/chainbridge-runtime-config.cere.dev.js]: Let's use `bridge-ui`

File: src/Contexts/Adaptors/SubstrateDestinationAdaptor.tsx [hunk 2]
```diff
@@ -0,0 +1,231 @@
+import React, { useCallback, useEffect, useState } from "react";
```
Comment (AndreiNavoichyk) [inline src/Contexts/Adaptors/SubstrateDestinationAdaptor.tsx:1]: Please, move class splitting to a separate PR. It does not relate to the feature.

File: src/Contexts/Adaptors/EVMAdaptors/EVMDestinationAdaptorProvider.tsx [hunk 3]
```diff
@@ -5,12 +5,19 @@ import { useEffect, useState } from "react";
...
   getСhainTransferFallbackConfig,
+  chainbridgeConfig,
 } from "../../../chainbridgeConfig";
 import { useNetworkManager } from "../../NetworkManagerContext";
 import { IDestinationBridgeProviderProps } from "../interfaces";
 import { DestinationBridgeContext } from "../../DestinationBridgeContext";
 import { getProvider, getErc20ProposalHash, VoteStatus } from "./helpers";
 import { Fallback } from "../../../Utils/Fallback";
+import { GA } from "../../../Utils/GA";
+const ga = new GA({
```
Comment (AndreiNavoichyk) [inline src/Contexts/Adaptors/EVMAdaptors/EVMDestinationAdaptorProvider.tsx]: Let's create an `AnalyticsService` and encapsulate all details (event name, parameters structure, etc.) there so in business code you need to trigger a method with required parameters.

File: src/Contexts/Adaptors/SubstrateDestinationAdaptor.tsx (hunk 2 above)
Comment (DmitriyMolch) [reply src/Contexts/Adaptors/SubstrateDestinationAdaptor.tsx:1]: Do we need to spend time on this? It's simpler to test everything in 1 PR so we spend less time on manual testing.

File: config/chainbridge-runtime-config.cere.dev.js (hunk 1 above)
Comment (Dmitriy1984Molchanov) [reply config/chainbridge-runtime-config.cere.dev.js:9]: This name already exists in line 6

File: src/Contexts/Adaptors/EVMAdaptors/EVMDestinationAdaptorProvider.tsx (hunk 3 above)
Comment (DmitriyMolch) [reply src/Contexts/Adaptors/EVMAdaptors/EVMDestinationAdaptorProvider.tsx]: This service was designed as a reusable utils library that can be used inside any application. So its encapsulation borders are correct in that terms. I didn't have a purpose to create some wrap around it and designed it to be universal and simple.

File: config/chainbridge-runtime-config.cere.dev.js (hunk 1 above)
Comment (DmitriyMolch) [reply config/chainbridge-runtime-config.cere.dev.js]: Decided to use `chainbridge-ui`

Comment (DmitriyMolch) [reply config/chainbridge-runtime-config.cere.dev.js]: fixed

File: src/Contexts/Adaptors/EVMAdaptors/EVMDestinationAdaptorProvider.tsx (hunk 3 above)
Comment (DmitriyMolch) [reply src/Contexts/Adaptors/EVMAdaptors/EVMDestinationAdaptorProvider.tsx]: fixed

File: CHANGELOG.md [hunk 4]
```diff
@@ -2,11 +2,8 @@
 
 ## vNext
 
-- Implemented fallback mechanism for Polygon, Ethereum to Cere, Cere to Polygon transfer flows to be able to update transaction status window in the case blockchain response didn't reach client.
-
-## v0.12.0
-
-- Updated Testnet/Mainnet urls
+- Implemented fallback mechanism for Polygon, Ethereum to Cere, Cere to Polygon transfer flows to be able to update transaction status window in the case blockchain response didn't reach client with GA integration
+- Updated Testnet urls
```
Comment (AndreiNavoichyk) [inline CHANGELOG.md]: Please, check the merge with `dev-cere`. `Update urls` has been already released in v0.12

File: src/Utils/GA.tsx [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+  private env: string;
+
+  constructor(options: Options) {
+    ReactGA.initialize(options.trackingId);
+    this.ga = ReactGA;
+    this.appName = options.appName;
+    this.env = options.env;
+  }
+
+  event(name: string, data: EventData) {
```
Comment (AndreiNavoichyk) [inline src/Utils/GA.tsx]: Each method should have a verb. Please, rename to `trackEvent` or `sendEvent`.

File: src/Services/Analytics.ts [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+};
+
+class AnalyticsService {
+  ga: GA;
+
+  constructor(options: Options) {
+    this.ga = new GA(options.ga);
+  }
+
+  transferIntransitEvent({
```
Comment (AndreiNavoichyk) [inline src/Services/Analytics.ts]: Each method should have a verb. Please, rename to `trackTransferIn**T**ransit`.

Apply the same for all methods.

File: src/Services/Analytics.ts [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+    address,
+    recipient,
+    nonce,
+    amount,
+  }: TransferEventData) {
+    this.ga.event("transfer_intransit", {
```
Comment (AndreiNavoichyk) [inline src/Services/Analytics.ts]: Add `_` after `in`

File: src/Services/Analytics.ts [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+  }: TransferEventData) {
+    this.ga.event("transfer_completed", {
+      address,
+      recipient,
+      nonce,
+      amount,
+    });
+  }
+
+  fallbackTransferCompletedEvent({
```
Comment (AndreiNavoichyk) [inline src/Services/Analytics.ts]: `trackTransferCompletedFromFallback`?
Please, apply to all fallback methods.

File: src/Services/Analytics.ts [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+    address,
+    recipient,
+    nonce,
+    amount,
+  }: TransferEventData) {
+    this.ga.event("fallback_transfer_completed", {
```
Comment (AndreiNavoichyk) [inline src/Services/Analytics.ts]: Let's use `fallback` as a suffix and not a prefix. For example, `transfer_completed_fallback`.
Please, apply to all fallback event names.

File: src/Contexts/Adaptors/SubstrateHomeAdaptor.tsx [hunk 10]
```diff
@@ -23,14 +13,21 @@ import {
...
 import { utils } from "ethers";
 import {
   SubstrateBridgeConfig,
-  getСhainTransferFallbackConfig,
+  chainbridgeConfig,
 } from "../../chainbridgeConfig";
 import { toFixedWithoutRounding } from "../../Utils/Helpers";
-import { Fallback } from "../../Utils/Fallback";
+import AnalyticsService from "../../Services/Analytics";
+const analytics = new AnalyticsService({
```
Comment (AndreiNavoichyk) [inline src/Contexts/Adaptors/SubstrateHomeAdaptor.tsx]: Can it be initialized once?

Comment (DmitriyMolch) [reply src/Contexts/Adaptors/SubstrateHomeAdaptor.tsx]: Theoretically yes,  we can try to create a separate  `AnalyticsProvider` or put it to the existing `NetworkManagerProvider`. What is more preferable?

File: src/Services/Analytics.ts (hunk 9 above)
Comment (DmitriyMolch) [reply src/Services/Analytics.ts]: In that case, the code becomes not readable as there is no meaning to the phrase "transfer completed fallback" as it doesn't shows finish`ed` event.  Actually not really important.

File: src/Services/Analytics.ts (hunk 6 above)
Comment (DmitriyMolch) [reply src/Services/Analytics.ts]: fixed

File: src/Services/Analytics.ts (hunk 7 above)
Comment (DmitriyMolch) [reply src/Services/Analytics.ts]: fixed

File: src/Services/Analytics.ts (hunk 8 above)
Comment (DmitriyMolch) [reply src/Services/Analytics.ts]: fixed

File: src/Services/Analytics.ts (hunk 9 above)
Comment (DmitriyMolch) [reply src/Services/Analytics.ts]: fixed

File: src/Contexts/Adaptors/SubstrateHomeAdaptor.tsx (hunk 10 above)
Comment (DmitriyMolch) [reply src/Contexts/Adaptors/SubstrateHomeAdaptor.tsx]: Implement it in `NetworkManagerProvider`

File: CHANGELOG.md (hunk 4 above)
Comment (DmitriyMolch) [reply CHANGELOG.md]: Fixed

File: src/Utils/GA.tsx (hunk 5 above)
Comment (DmitriyMolch) [reply src/Utils/GA.tsx]: fixed

File: src/Services/Analytics.ts (hunk 9 above)
Comment (AndreiNavoichyk) [reply src/Services/Analytics.ts]: Feel free to make it more readable by updating to smth like `transfer_completed_from_fallback`.

**Review by AndreiNavoichyk — APPROVED** [review]

File: src/Services/Analytics.ts (hunk 9 above)
Comment (DmitriyMolch) [reply src/Services/Analytics.ts]: fixed

**Review by AndreiNavoichyk — APPROVED** [review]
